### PR TITLE
Add timeout for membership probe_uri

### PR DESCRIPTION
**Pull request by vanyarock01**
What has been done? Why? What problem is being solved?

I didn't forget about

- [x] Tests
- [x] Changelog
- [ ] Documentation

Close #???

### CODE REVIEW

File: cartridge.lua [hunk 1]
```diff
@@ -216,6 +217,8 @@ local function cfg(opts, box_opts)
         console_sock = '?string',
         webui_blacklist = '?table',
         upgrade_schema = '?boolean',
+        probe_uri_timeout = '?number',
+        probe_uri_delay = '?number',
```
Comment (rosik) [inline cartridge.lua:221]: Options variety increase complexity. I would recommend acquiring it with `argparse` and keeping it private.

File: cartridge.lua [hunk 2]
```diff
@@ -461,9 +464,23 @@ local function cfg(opts, box_opts)
...
+        if ok then
+            resolved = true
+            break
+        end
+        log.info('Can not ping myself: %s', estr)
+        time = time - delay
+    end
+
+    if not resolved then
+        return nil, CartridgeCfgError:new('Instance has not been resolved')
```
Comment (rosik) [inline cartridge.lua:483]: This loop is incorrect and too verbose.

```
while true do 
  local ok, err = ...
  if ok then
    break
  elseif time > deadline then
    return nil, err
  else 
    fiber.sleep(pause)
  end
end
```

File: cartridge.lua [hunk 3]
```diff
@@ -461,9 +464,23 @@ local function cfg(opts, box_opts)
...
+    local resolved = false
+
+    while time > 0 do
+        local ok, estr = membership.probe_uri(membership.myself().uri)
+        if ok then
+            resolved = true
+            break
+        end
+        log.info('Can not ping myself: %s', estr)
+        time = time - delay
```
Comment (rosik) [inline cartridge.lua:479]: What is the purpose of delay?

Comment (vanyarock01) [reply cartridge.lua:479]: This is the delay between checks

File: test/unit/cfg_errors_test.lua [hunk 4]
```diff
@@ -263,6 +278,71 @@ g.test_console_sock = function()
...
+
+    -- default probe_uri_timeout: 1.0s
+    local ok, err = cartridge.cfg({
+        workdir = g.tempdir,
+        advertise_uri = 'localhost:13001',
+        roles = {},
+    })
+    t.assert(ok, err)
+
+    stop_http_service()
```
Comment (rosik) [inline test/unit/cfg_errors_test.lua:316]: better use `http_enabled = false`

File: test/unit/cfg_errors_test.lua [hunk 5]
```diff
@@ -263,6 +278,71 @@ g.test_console_sock = function()
...
+        fiber.sleep(0.3)
+        dns_resolved = true
+    end)
+
+    -- default probe_uri_timeout: 1.0s
+    local ok, err = cartridge.cfg({
+        workdir = g.tempdir,
+        advertise_uri = 'localhost:13001',
+        roles = {},
+    })
```
Comment (rosik) [inline test/unit/cfg_errors_test.lua:313]: You were going to make default 0. I suggest passing timeout explicitly with `os.setenv`

File: cartridge.lua [hunk 6]
```diff
@@ -461,9 +462,27 @@ local function cfg(opts, box_opts)
...
+    end
+
+    local delay = require('membership.options').ACK_TIMEOUT_SECONDS
+    local deadline = fiber.time() + (probe_uri_opts.probe_uri_timeout or 0)
+
+    while true do
+        local next_wakeup = fiber.time() + delay 
+        local ok, estr = membership.probe_uri(membership.myself().uri)
+        local now = fiber.time()
+        if ok then
```
Comment (vasiliy-t) [inline cartridge.lua:478]: consider adding log.verbose("it's ok .....") could be useful in case of debug

File: cartridge.lua [hunk 7]
```diff
@@ -461,9 +462,27 @@ local function cfg(opts, box_opts)
...
+            break
+        elseif now >= deadline then
+            return nil, CartridgeCfgError:new('Can not ping myself: %s', estr)
+        else
+            log.info('Can not ping myself: %s', estr)
```
Comment (vasiliy-t) [inline cartridge.lua:483]: consider logging remaining time

File: cartridge.lua [hunk 8]
```diff
@@ -461,9 +462,27 @@ local function cfg(opts, box_opts)
...
+    local deadline = fiber.time() + (probe_uri_opts.probe_uri_timeout or 0)
+
+    while true do
+        local next_wakeup = fiber.time() + delay 
+        local ok, estr = membership.probe_uri(membership.myself().uri)
+        local now = fiber.time()
+        if ok then
+            break
+        elseif now >= deadline then
+            return nil, CartridgeCfgError:new('Can not ping myself: %s', estr)
```
Comment (vasiliy-t) [inline cartridge.lua:481]: "can not ping myself" because of what? consider adding something like "deadline exceeded" to log message

Comment (rosik) [reply cartridge.lua:481]: because of `%s, estr`

File: CHANGELOG.rst [hunk 9]
```diff
@@ -22,6 +22,7 @@ Added
   failure, a user could either revert it (promote previous leader), or force
   promotion to be inconsistent.
 - Early logger initialization (for Tarantool > 2.5.0-100, which supports it).
+- Probe uri timeout in cartridge.cfg.
```
Comment (rosik) [inline CHANGELOG.rst:25]:
```suggestion
- Add `probe_uri_timeout` argparse option responsible for retrying
  "Can't ping myself" error on startup.
```

**Pull request merged**